In [ ]:
import numpy as np
import pandas as pd
from pathlib import Path
from tqdm import tqdm
from floweaver import *
from numpy import linalg
from scipy import sparse
import xlsxwriter

In [ ]:
from unidecode import unidecode
import re

In [ ]:
# Import data
df_customer = pd.read_excel(f"Vòng 1-Working.xlsx", sheet_name = "customer")
df_ticket_table = pd.read_excel(f"Vòng 1-Working.xlsx", sheet_name = "ticket")
df_film = pd.read_excel(f"Vòng 1-Working.xlsx", sheet_name = "film")


## 1. Customer table

In [ ]:
df_customer

customerid    DOB gender                           address  \
0          0000000014  36268     Nữ            hoa khe quan thanh khe   
1          0000000034  35901    Nam                          man thai   
2          0000000051  34319    Nam                 que son quang nam   
3          0000000081  36472    Nam                      truong chinh   
4          0000000098  34608    Nam                            cẩm lệ   
...               ...    ...    ...                               ...   
4474        KH9855766  34885     Nữ  A, THUAN PHUOC,HAI CHAU, DA NANG   
4475        KH9958204  36892     Nữ                      Phước Mỹ, ĐN   
4476  WEBS00000043900  32311     Nữ                 nguyen gian thanh   
4477  WEBS00000044909  35316     Nữ                      ngũ hành sơn   
4478  WEBS00000046015  37372     Nữ         Thanh Khê Tây - Thanh Khê   

              Website          job        industry  
0         KH0104|0345      student        computer  
1         KH0104|0333      student  health service  
2         KH0104|0255  blue collar       economics  
3         KH0104|0293      student       economics  
4        KH0104|40580  blue collar  health service  
...               ...          ...             ...  
4474              NaN   specialist         finance  
4475              NaN     teenager             NaN  
4476  KHWEBS|00006890   specialist    construction  
4477  KHWEBS|00007899  blue collar         finance  
4478  KHWEBS|00009005     teenager             NaN  

[4479 rows x 7 columns]

In [ ]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4479 entries, 0 to 4478
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerid  4479 non-null   object
 1   DOB         4479 non-null   int64 
 2   gender      4479 non-null   object
 3   address     4478 non-null   object
 4   Website     3943 non-null   object
 5   job         4479 non-null   object
 6   industry    3357 non-null   object
dtypes: int64(1), object(6)
memory usage: 245.1+ KB


Mission:
- Redefine DOB data to the correct format
-

In [ ]:
for i in range(len(df_customer['address'])):
  try:
    df_customer['address'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_customer['address'][i])).lower())
  except:
    df_customer['address'][i] = 0

/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/1262063627.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_customer['address'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_customer['address'][i])).lower())
/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/1262063627.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_customer['address'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_customer['address'][i])).lower())
/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/1262063627.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from

In [ ]:
# To trim the spacing at the begining
df_customer['address'].str.lstrip(' ')
df_customer

customerid    DOB gender                         address  \
0          0000000014  36268     Nữ          hoa khe quan thanh khe   
1          0000000034  35901    Nam                        man thai   
2          0000000051  34319    Nam               que son quang nam   
3          0000000081  36472    Nam                    truong chinh   
4          0000000098  34608    Nam                          cam le   
...               ...    ...    ...                             ...   
4474        KH9855766  34885     Nữ  a thuan phuoc hai chau da nang   
4475        KH9958204  36892     Nữ                     phuoc my dn   
4476  WEBS00000043900  32311     Nữ               nguyen gian thanh   
4477  WEBS00000044909  35316     Nữ                    ngu hanh son   
4478  WEBS00000046015  37372     Nữ         thanh khe tay thanh khe   

              Website          job        industry  
0         KH0104|0345      student        computer  
1         KH0104|0333      student  health service  
2         KH0104|0255  blue collar       economics  
3         KH0104|0293      student       economics  
4        KH0104|40580  blue collar  health service  
...               ...          ...             ...  
4474              NaN   specialist         finance  
4475              NaN     teenager             NaN  
4476  KHWEBS|00006890   specialist    construction  
4477  KHWEBS|00007899  blue collar         finance  
4478  KHWEBS|00009005     teenager             NaN  

[4479 rows x 7 columns]

In [ ]:
# # Export to excel
# df_customer.to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/customer_cleaned.xlsx", sheet_name="customer", engine="xlsxwriter")

In [ ]:
df_customer['customerid'].nunique()

4479

In [ ]:
(df_customer['Website'].isna()).sum()

536

In [ ]:
df_customer["job"].nunique()

5

In [ ]:
df_customer["industry"].isna().sum()

1122

In [ ]:
df_customer["address"]

0               hoa khe quan thanh khe
1                             man thai
2                    que son quang nam
3                         truong chinh
4                               cam le
                     ...              
4474    a thuan phuoc hai chau da nang
4475                       phuoc my dn
4476                 nguyen gian thanh
4477                      ngu hanh son
4478           thanh khe tay thanh khe
Name: address, Length: 4479, dtype: object

Thông tin adress có khá nhiều lỗi
- Có cả thông tin full tiếng Việt và thông tin không dấu
- Lẫn lộn chữ hoa, thường
- Dư spacing
- Có người điền cả quận, có người chỉ điền đường
- Với thông tin có đường, phường, quận, có người điền với dấu phẩy (ngăn cách), có người lại không

As the longest length is up to 69 characters (including spacings) we will separate them to a new dataframe and working with them

In [ ]:
df_customer

customerid    DOB gender                         address  \
0          0000000014  36268     Nữ          hoa khe quan thanh khe   
1          0000000034  35901    Nam                        man thai   
2          0000000051  34319    Nam               que son quang nam   
3          0000000081  36472    Nam                    truong chinh   
4          0000000098  34608    Nam                          cam le   
...               ...    ...    ...                             ...   
4474        KH9855766  34885     Nữ  a thuan phuoc hai chau da nang   
4475        KH9958204  36892     Nữ                     phuoc my dn   
4476  WEBS00000043900  32311     Nữ               nguyen gian thanh   
4477  WEBS00000044909  35316     Nữ                    ngu hanh son   
4478  WEBS00000046015  37372     Nữ         thanh khe tay thanh khe   

              Website          job        industry  
0         KH0104|0345      student        computer  
1         KH0104|0333      student  health service  
2         KH0104|0255  blue collar       economics  
3         KH0104|0293      student       economics  
4        KH0104|40580  blue collar  health service  
...               ...          ...             ...  
4474              NaN   specialist         finance  
4475              NaN     teenager             NaN  
4476  KHWEBS|00006890   specialist    construction  
4477  KHWEBS|00007899  blue collar         finance  
4478  KHWEBS|00009005     teenager             NaN  

[4479 rows x 7 columns]

In [ ]:
df_customer['address']

0               hoa khe quan thanh khe
1                             man thai
2                    que son quang nam
3                         truong chinh
4                               cam le
                     ...              
4474    a thuan phuoc hai chau da nang
4475                       phuoc my dn
4476                 nguyen gian thanh
4477                      ngu hanh son
4478           thanh khe tay thanh khe
Name: address, Length: 4479, dtype: object

In [ ]:
# Get streetname dataframe
# Main resource: https://congdulieu.vn/dulieuchitiet/1037328
df_dn_street = pd.read_excel("danang_street_name.xlsx", sheet_name="street")

# Subset
df_dn_street = df_dn_street[["Tên đường", "Từ khóa"]]
# Rename
df_dn_street = df_dn_street.rename(
    columns={"Tên đường": "street_name", "Từ khóa": "keyword"}
)

# Clean
for i in range(len(df_dn_street['street_name'])):
  try:
    df_dn_street['street_name'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_dn_street['street_name'][i])).lower())
  except:
    df_dn_street['street_name'][i] = 0

df_dn_street

street_name                            keyword
0                3 2                3 tháng 2, Hải Châu
1               30 4               30 tháng 4, Hải Châu
2                2 9  2 tháng 9, Hải Châu, quốc khánh, 
3     phan dinh giot         Phan Đinh Giót, liên chiểu
4        ly te xuyen                        Lý Tế Xuyên
...              ...                                ...
1960       an ngai 5                                NaN
1961       an ngai 6                                NaN
1962       an ngai 7                                NaN
1963       an ngai 8                                NaN
1964       an ngai 9                                NaN

[1965 rows x 2 columns]

In [ ]:
# Danh sách đường ở Đà Nẵng: https://moovitapp.com/index/vi/ph%C6%B0%C6%A1ng_ti%E1%BB%87n_c%C3%B4ng_c%E1%BB%99ng-streets-Da_Nang-1-6123
# Or here: https://dangbodanang.vn/vankien-tulieu/tulieu/tulieu/tenduongphodanangnguonhoikhoahoclichsudanang/54/10/59

# Cần web scrapping đống đường này xuống một file
# https://congdulieu.vn/dulieuchitiet/1037328

In [ ]:
# Tham khảo: https://github.com/ThangLeQuoc/vietnamese-provinces-database/blob/master/README_vi.md

In [ ]:
# Get the dataframe of district and ward name
df_dn_district = pd.read_excel("danang_street_name.xlsx", sheet_name="district, ward")

# Rename
df_dn_district = df_dn_district.rename(
    columns={"Quận Huyện": "district", "Phường Xã": "ward"}
)

# Clean for district column
for i in range(len(df_dn_district['district'])):
  try:
    df_dn_district['district'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_dn_district['district'][i])).lower())
  except:
    df_dn_district['district'][i] = 0

# Clean for ward column
for i in range(len(df_dn_district['ward'])):
  try:
    df_dn_district['ward'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_dn_district['ward'][i])).lower())
  except:
    df_dn_district['ward'][i] = df_dn_district['district'][i]

df_dn_district

/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/3931264993.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dn_district['district'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_dn_district['district'][i])).lower())
/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/3931264993.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dn_district['district'][i] = re.sub('[^A-Za-z0-9]+', ' ', (unidecode(df_dn_district['district'][i])).lower())
/var/folders/v2/2rzp2kqj0rs9_jtdfnc5d55w0000gn/T/ipykernel_20909/3931264993.py:12: SettingWithCopyWarning: 
A value is trying to be set on a c

STT           district  Mã QH                   ward    Mã PX     Cấp
0     1    quan lien chieu    490    phuong hoa hiep bac  20194.0  Phường
1     2    quan lien chieu    490    phuong hoa hiep nam  20195.0  Phường
2     3    quan lien chieu    490   phuong hoa khanh bac  20197.0  Phường
3     4    quan lien chieu    490   phuong hoa khanh nam  20198.0  Phường
4     5    quan lien chieu    490        phuong hoa minh  20200.0  Phường
5     6     quan thanh khe    491       phuong tam thuan  20203.0  Phường
6     7     quan thanh khe    491   phuong thanh khe tay  20206.0  Phường
7     8     quan thanh khe    491  phuong thanh khe dong  20207.0  Phường
8     9     quan thanh khe    491         phuong xuan ha  20209.0  Phường
9    10     quan thanh khe    491       phuong tan chinh  20212.0  Phường
10   11     quan thanh khe    491      phuong chinh gian  20215.0  Phường
11   12     quan thanh khe    491      phuong vinh trung  20218.0  Phường
12   13     quan thanh khe    491       phuong thac gian  20221.0  Phường
13   14     quan thanh khe    491          phuong an khe  20224.0  Phường
14   15     quan thanh khe    491         phuong hoa khe  20225.0  Phường
15   16      quan hai chau    492      phuong thanh binh  20227.0  Phường
16   17      quan hai chau    492     phuong thuan phuoc  20230.0  Phường
17   18      quan hai chau    492     phuong thach thang  20233.0  Phường
18   19      quan hai chau    492      phuong hai chau i  20236.0  Phường
19   20      quan hai chau    492     phuong hai chau ii  20239.0  Phường
20   21      quan hai chau    492      phuong phuoc ninh  20242.0  Phường
21   22      quan hai chau    492   phuong hoa thuan tay  20245.0  Phường
22   23      quan hai chau    492  phuong hoa thuan dong  20246.0  Phường
23   24      quan hai chau    492       phuong nam duong  20248.0  Phường
24   25      quan hai chau    492       phuong binh hien  20251.0  Phường
25   26      quan hai chau    492      phuong binh thuan  20254.0  Phường
26   27      quan hai chau    492   phuong hoa cuong bac  20257.0  Phường
27   28      quan hai chau    492   phuong hoa cuong nam  20258.0  Phường
28   29       quan son tra    493       phuong tho quang  20263.0  Phường
29   30       quan son tra    493   phuong nai hien dong  20266.0  Phường
30   31       quan son tra    493        phuong man thai  20269.0  Phường
31   32       quan son tra    493      phuong an hai bac  20272.0  Phường
32   33       quan son tra    493        phuong phuoc my  20275.0  Phường
33   34       quan son tra    493      phuong an hai tay  20278.0  Phường
34   35       quan son tra    493     phuong an hai dong  20281.0  Phường
35   36  quan ngu hanh son    494           phuong my an  20284.0  Phường
36   37  quan ngu hanh son    494         phuong khue my  20285.0  Phường
37   38  quan ngu hanh son    494         phuong hoa quy  20287.0  Phường
38   39  quan ngu hanh son    494         phuong hoa hai  20290.0  Phường
39   40        quan cam le    495      phuong khue trung  20260.0  Phường
40   41        quan cam le    495        phuong hoa phat  20305.0  Phường
41   42        quan cam le    495          phuong hoa an  20306.0  Phường
42   43        quan cam le    495     phuong hoa tho tay  20311.0  Phường
43   44        quan cam le    495    phuong hoa tho dong  20312.0  Phường
44   45        quan cam le    495        phuong hoa xuan  20314.0  Phường
45   46     huyen hoa vang    497             xa hoa bac  20293.0      Xã
46   47     huyen hoa vang    497            xa hoa lien  20296.0      Xã
47   48     huyen hoa vang    497            xa hoa ninh  20299.0      Xã
48   49     huyen hoa vang    497             xa hoa son  20302.0      Xã
49   50     huyen hoa vang    497            xa hoa nhon  20308.0      Xã
50   51     huyen hoa vang    497             xa hoa phu  20317.0      Xã
51   52     huyen hoa vang    497           xa hoa phong  20320.0      Xã
52   53     huyen hoa vang    497            xa hoa chau  20323.0      Xã
53   54 

In [ ]:
# Now, get the list of district name and strim a part of unuse information
df_dn_district["district_new"] = df_dn_district["district"].map(
    lambda x: x.lstrip("quan|huyen")
)
df_dn_district["district_new"] = df_dn_district["district_new"].str.lstrip(" ")
# Also do it with ward
df_dn_district["ward_new"] = df_dn_district["ward"].map(
    lambda x: x.lstrip("huyen|phuong|xa")
)
df_dn_district["ward_new"] = df_dn_district["ward_new"].str.lstrip(" ")
# Get a list of unique district and ward
list_dn_district = list(df_dn_district["district_new"].unique())
list_dn_ward = list(df_dn_district["ward_new"].unique())
list_dn_ward.append('hoa khanh')

In [ ]:
list_dn_ward

['hoa hiep bac',
 'hoa hiep nam',
 'hoa khanh bac',
 'hoa khanh nam',
 'hoa minh',
 'tam thuan',
 'thanh khe tay',
 'thanh khe dong',
 'xuan ha',
 'tan chinh',
 'chinh gian',
 'vinh trung',
 'thac gian',
 'an khe',
 'hoa khe',
 'thanh binh',
 'thuan phuoc',
 'thach thang',
 'hai chau i',
 'hai chau ii',
 'phuoc ninh',
 'hoa thuan tay',
 'hoa thuan dong',
 'nam duong',
 'binh hien',
 'binh thuan',
 'hoa cuong bac',
 'hoa cuong nam',
 'tho quang',
 'nai hien dong',
 'man thai',
 'an hai bac',
 'phuoc my',
 'an hai tay',
 'an hai dong',
 'my an',
 'khue my',
 'hoa quy',
 'hoa hai',
 'khue trung',
 'hoa phat',
 'hoa an',
 'hoa tho tay',
 'hoa tho dong',
 'hoa xuan',
 'hoa bac',
 'hoa lien',
 'hoa ninh',
 'hoa son',
 'hoa nhon',
 'hoa phu',
 'hoa phong',
 'hoa chau',
 'hoa tien',
 'hoa phuoc',
 'hoa khuong',
 'hoang sa',
 'hoa khanh']

In [ ]:
# Get a list of streetname
df_dn_street['street_name'] = df_dn_street['street_name'].str.lstrip(" ")
df_dn_street['street_name'] = df_dn_street['street_name'].str.rstrip(" ")
list_dn_street = list(df_dn_street['street_name'].unique())

In [ ]:
# The first code line is to check whether the address contains "dn" or "da nang"
df_customer['contains_dn'] = df_customer['address'].str.contains(('dn|da nang'))

# Then test with district list and ward list
df_customer['contains_dt'] = df_customer['address'].str.contains('|'.join(list_dn_district))
df_customer['contains_w'] = df_customer['address'].str.contains('|'.join(list_dn_ward))

# Finally, test with streetname
df_customer['contains_str'] = df_customer['address'].str.contains('|'.join(list_dn_street))


In [ ]:
# Attract the subset with all True value with "dn" or "da nang"
df_customer_dn = df_customer[(df_customer['contains_dn']) | (df_customer['contains_dt']) | (df_customer['contains_w']) | (df_customer['contains_str'])]

# Get a dataframe of not in dn
df_customer_notdn = df_customer[(df_customer['contains_dn'] == False) & (df_customer['contains_dt'] == False) & (df_customer['contains_w'] == False) &( df_customer['contains_str'] == False)]

In [ ]:
# Get a dataframe of not in dn
df_customer_notdn = df_customer[(df_customer['contains_dn'] == False) & (df_customer['contains_dt'] == False) & (df_customer['contains_w'] == False) &( df_customer['contains_str'] == False)]

In [ ]:
# Create a column is danang which "yes" for all value in df_customer_dn and no for value in df_customer_notdn
# Checking conditions for 'is_danang'
conditions_dn = (df_customer['contains_dn'] == True) | (df_customer['contains_dt'] == True) | (df_customer['contains_w'] == True) | (df_customer['contains_str'] == True)
conditions_notdn = ~conditions_dn

# Creating 'is_danang' column
df_customer['is_danang'] = 'no'
df_customer.loc[conditions_dn, 'is_danang'] = 'yes'

In [ ]:
(df_customer["is_danang"]=="no").value_counts()

is_danang
False    4258
True      221
Name: count, dtype: int64

In [ ]:
# Drop some unnecessary colummns
df_customer_1 = df_customer.drop(columns=["contains_dn", "contains_dt", "contains_w", "contains_str"])

The above df_customer_notdn has 221 rows, which has:
- The address in Danang city but has typo
- The streetname is not in the given list (maybe new streetname?)
- Not in Danang city


In [ ]:
df_customer_1.to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/customer_cleanned_2.xlsx", index=False, engine='xlsxwriter')

In [ ]:
#export to excel
def export_to_excel(file_name, df1, df2):
    # Create an ExcelWriter object to write to the Excel file
    with pd.ExcelWriter(file_name, engine='xlsxwriter') as writer:
        # Write each DataFrame to a separate sheet
        df1.to_excel(writer, sheet_name='street', index=False)
        df2.to_excel(writer, sheet_name='district-ward', index=False)

export_to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/dn-street-clean.xlsx",  df_dn_street, df_dn_district)

In [ ]:
# # Create a function to check for similarity and assign "yes" or "no" based on conditions
# def check_similarity(row):
#     address = row['address']
#     street_name = row['street_name']
#     keyword = row['keyword']

#     # Check if a part of the string in address appears in street_name
#     if address in street_name or street_name in address:
#         return 'yes'
#     else:
#         return 'no'

# # Apply the function to create a new column in the first dataframe
# df_customer_add['similarity_category'] = df_customer_add.apply(check_similarity, axis=1)

# # Display the updated dataframe
# print(df_customer_add)

In [ ]:
# export_to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/dn-clean-2.xlsx",  df_customer_dn, df_customer_notdn)

## 2. Ticket table

In [ ]:
df_ticket_table

orderid cashier            saledate     total  \
0      10006052019B0225  emp002 2019-05-06 16:40:43   90000.0   
1      10006052019B0225  emp002 2019-05-06 16:40:43   90000.0   
2      10022052019B0167  emp011 2019-05-22 20:30:28   90000.0   
3      10022052019B0167  emp011 2019-05-22 20:30:28   90000.0   
4      10130052019B0141  emp005 2019-05-30 19:28:44   90000.0   
...                 ...     ...                 ...       ...   
35469  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35470  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35471  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35472  12508052019B0036  emp007 2019-05-08 19:26:47   90000.0   
35473  12508052019B0036  emp007 2019-05-08 19:26:47   90000.0   

            customerid  ticketcode       date      time slot  room  \
0           0000000014    20074925 2019-05-06  18:15:00  G04     3   
1           0000000014    20074924 2019-05-06  18:15:00  G03     3   
2           0000000034    20097042 2019-05-22  20:45:00  E04     2   
3           0000000034    20097041 2019-05-22  20:45:00  E03     2   
4           0000000034    20106761 2019-05-30  19:30:00  F04     1   
...                ...         ...        ...       ...  ...   ...   
35469  WEBS00000044909    20107072 2019-05-30  21:35:00  B10     4   
35470  WEBS00000044909    20107071 2019-05-30  21:35:00  B09     4   
35471  WEBS00000044909    20107069 2019-05-30  21:35:00  B07     4   
35472  WEBS00000046015    20077797 2019-05-08  19:15:00  C14     4   
35473  WEBS00000046015    20077796 2019-05-08  19:15:00  C13     4   

                                          film slot type ticket type  \
0                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
1                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
2      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
3      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
4                       NGÔI ĐỀN KỲ QUÁI (C18)       ĐƠN  Thành viên   
...                                        ...       ...         ...   
35469  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35470  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35471  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35472                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
35473                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   

       ticket price popcorn  
0             45000   Không  
1             45000   Không  
2             45000   Không  
3             45000   Không  
4             45000   Không  
...             ...     ...  
35469         45000   Không  
35470         45000   Không  
35471         45000   Không  
35472         45000   Không  
35473         45000   Không  

[35474 rows x 15 columns]

In [ ]:
df_ticket = df_ticket_table.copy()

In [ ]:
# Rename date to showdate
df_ticket = df_ticket.rename(columns = {"date":"showdate"})

In [ ]:
# Fix the format of saledate and date
# df_ticket['saledate'] = pd.to_datetime(df_ticket["saledate"].dt.strftime('%m-%Y'))
# df_ticket['saledate'] = pd.to_datetime(df_ticket["saledate"], dayfirst=True)
df_ticket

orderid cashier            saledate     total  \
0      10006052019B0225  emp002 2019-05-06 16:40:43   90000.0   
1      10006052019B0225  emp002 2019-05-06 16:40:43   90000.0   
2      10022052019B0167  emp011 2019-05-22 20:30:28   90000.0   
3      10022052019B0167  emp011 2019-05-22 20:30:28   90000.0   
4      10130052019B0141  emp005 2019-05-30 19:28:44   90000.0   
...                 ...     ...                 ...       ...   
35469  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35470  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35471  10130052019B0192  emp005 2019-05-30 21:49:46  180000.0   
35472  12508052019B0036  emp007 2019-05-08 19:26:47   90000.0   
35473  12508052019B0036  emp007 2019-05-08 19:26:47   90000.0   

            customerid  ticketcode   showdate      time slot  room  \
0           0000000014    20074925 2019-05-06  18:15:00  G04     3   
1           0000000014    20074924 2019-05-06  18:15:00  G03     3   
2           0000000034    20097042 2019-05-22  20:45:00  E04     2   
3           0000000034    20097041 2019-05-22  20:45:00  E03     2   
4           0000000034    20106761 2019-05-30  19:30:00  F04     1   
...                ...         ...        ...       ...  ...   ...   
35469  WEBS00000044909    20107072 2019-05-30  21:35:00  B10     4   
35470  WEBS00000044909    20107071 2019-05-30  21:35:00  B09     4   
35471  WEBS00000044909    20107069 2019-05-30  21:35:00  B07     4   
35472  WEBS00000046015    20077797 2019-05-08  19:15:00  C14     4   
35473  WEBS00000046015    20077796 2019-05-08  19:15:00  C13     4   

                                          film slot type ticket type  \
0                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
1                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
2      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
3      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
4                       NGÔI ĐỀN KỲ QUÁI (C18)       ĐƠN  Thành viên   
...                                        ...       ...         ...   
35469  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35470  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35471  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35472                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
35473                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   

       ticket price popcorn  
0             45000   Không  
1             45000   Không  
2             45000   Không  
3             45000   Không  
4             45000   Không  
...             ...     ...  
35469         45000   Không  
35470         45000   Không  
35471         45000   Không  
35472         45000   Không  
35473         45000   Không  

[35474 rows x 15 columns]

In [ ]:
df_ticket["orderid"].isna().sum()

96

There's about 96 data in the ticket["orderid"] are NaN/NA value --> Order from web
Solution: Assign new orderid for web.
There're some tendency of people orderred from website:
- Same cashier (which is Website)
- Same customerid
- Same film
- Same date
- Same time
- Same room

In [ ]:
# Get the different of saledate and showdate
df_ticket.loc[df_ticket["cashier"]!="Website", "saledate-showdate"] = df_ticket["showdate"]-df_ticket["saledate"]
df_ticket["saledate-showdate"] = df_ticket["saledate-showdate"].dt.days
# Calculate the average difference
print("Thời gian trung bình từ ngày đặt vé đến ngày xem phim: ", round(df_ticket["saledate-showdate"].mean()), " ngày")

Thời gian trung bình từ ngày đặt vé đến ngày xem phim:  -1  ngày


In [ ]:
# Append the average date different to cashier is website
df_ticket.loc[df_ticket["cashier"]=="Website", "saledate-showdate"] = round(df_ticket["saledate-showdate"].mean())

In [ ]:
df_ticket_1 = df_ticket.copy()

In [ ]:
# Assign the order id for web order
# Giả sử, với same customerid và same showdate, người mua vé mua cùng một orderid
df_web_1 = df_ticket_1[["customerid", "showdate"]]
df_web_1

customerid   showdate
0           0000000014 2019-05-06
1           0000000014 2019-05-06
2           0000000034 2019-05-22
3           0000000034 2019-05-22
4           0000000034 2019-05-30
...                ...        ...
35469  WEBS00000044909 2019-05-30
35470  WEBS00000044909 2019-05-30
35471  WEBS00000044909 2019-05-30
35472  WEBS00000046015 2019-05-08
35473  WEBS00000046015 2019-05-08

[35474 rows x 2 columns]

In [ ]:
# Get the new orderid for people order on website
df_ticket_1.loc[(df_ticket_1["cashier"] == "Website") & (df_ticket_1["orderid"].isnull()), "orderid"] = "website" + df_ticket_1["customerid"].astype(str) + df_ticket_1["showdate"].astype(str)

# Assume saledate is 15 days before the showdate
# df_ticket_1["saledate"] = df_ticket_1["showdate"] - pd.to_timedelta(15, unit = "D")
df_ticket_1["saledate"] = df_ticket_1["showdate"]

In [ ]:
# Append the No popcorn to cashier is website
df_ticket_1.loc[df_ticket_1["cashier"]=="Website", "popcorn"] = "Không"

In [ ]:
df_ticket_1

orderid cashier   saledate     total       customerid  \
0      10006052019B0225  emp002 2019-05-06   90000.0       0000000014   
1      10006052019B0225  emp002 2019-05-06   90000.0       0000000014   
2      10022052019B0167  emp011 2019-05-22   90000.0       0000000034   
3      10022052019B0167  emp011 2019-05-22   90000.0       0000000034   
4      10130052019B0141  emp005 2019-05-30   90000.0       0000000034   
...                 ...     ...        ...       ...              ...   
35469  10130052019B0192  emp005 2019-05-30  180000.0  WEBS00000044909   
35470  10130052019B0192  emp005 2019-05-30  180000.0  WEBS00000044909   
35471  10130052019B0192  emp005 2019-05-30  180000.0  WEBS00000044909   
35472  12508052019B0036  emp007 2019-05-08   90000.0  WEBS00000046015   
35473  12508052019B0036  emp007 2019-05-08   90000.0  WEBS00000046015   

       ticketcode   showdate      time slot  room  \
0        20074925 2019-05-06  18:15:00  G04     3   
1        20074924 2019-05-06  18:15:00  G03     3   
2        20097042 2019-05-22  20:45:00  E04     2   
3        20097041 2019-05-22  20:45:00  E03     2   
4        20106761 2019-05-30  19:30:00  F04     1   
...           ...        ...       ...  ...   ...   
35469    20107072 2019-05-30  21:35:00  B10     4   
35470    20107071 2019-05-30  21:35:00  B09     4   
35471    20107069 2019-05-30  21:35:00  B07     4   
35472    20077797 2019-05-08  19:15:00  C14     4   
35473    20077796 2019-05-08  19:15:00  C13     4   

                                          film slot type ticket type  \
0                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
1                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
2      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
3      JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
4                       NGÔI ĐỀN KỲ QUÁI (C18)       ĐƠN  Thành viên   
...                                        ...       ...         ...   
35469  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35470  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35471  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   
35472                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
35473                        AVENGERS: HỒI KẾT       ĐƠN  Thành viên   

       ticket price popcorn  saledate-showdate  
0             45000   Không               -1.0  
1             45000   Không               -1.0  
2             45000   Không               -1.0  
3             45000   Không               -1.0  
4             45000   Không               -1.0  
...             ...     ...                ...  
35469         45000   Không               -1.0  
35470         45000   Không               -1.0  
35471         45000   Không               -1.0  
35472         45000   Không               -1.0  
35473         45000   Không               -1.0  

[35474 rows x 16 columns]

In [ ]:
df_ticket_1["ticketcount"] = df_ticket_1.groupby('orderid')["ticketcode"].transform("count")

In [ ]:
df_ticket_1.loc[(df_ticket_1["cashier"] == "Website") & (df_ticket_1["total"].isnull()), "total"] = df_ticket_1["ticket price"]*df_ticket_1["ticketcount"]
df_ticket_1.loc[(df_ticket_1["total"]==0), "total"] = df_ticket_1["ticket price"]*df_ticket_1["ticketcount"]

In [ ]:
# df_ticket_1[df_ticket_1["saledate-showdate"]==0]

In [ ]:
# df_ticket_1.to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/ticket_cleanned_2.xlsx", index=False, engine='xlsxwriter')

In [ ]:
df_ticket_distinct = df_ticket_1.groupby('orderid').first().reset_index()

In [ ]:
df_ticket_distinct

orderid  cashier   saledate     total  customerid  \
0                 10001052019B0003   emp008 2019-05-01  160000.0  0000024994   
1                 10001052019B0004   emp008 2019-05-01   75000.0  0000024994   
2                 10001052019B0005   emp008 2019-05-01  225000.0   KH6166700   
3                 10001052019B0006   emp008 2019-05-01  150000.0   KH6166700   
4                 10001052019B0007   emp008 2019-05-01  160000.0   KH6166700   
...                            ...      ...        ...       ...         ...   
14636  website00011910712019-05-25  Website 2019-05-25  135000.0  0001191071   
14637  website00011910712019-05-26  Website 2019-05-26   90000.0  0001191071   
14638  website00030041612019-05-25  Website 2019-05-25   90000.0  0003004161   
14639  website00030086172019-05-17  Website 2019-05-17  135000.0  0003008617   
14640  website00030086172019-05-30  Website 2019-05-30   90000.0  0003008617   

       ticketcode   showdate      time slot  room  \
0        20068404 2019-05-01  07:30:00  H05     2   
1        20068405 2019-05-01  07:30:00  E10     2   
2        20068406 2019-05-01  08:45:00  G11     1   
3        20068409 2019-05-01  07:30:00  E06     2   
4        20068411 2019-05-01  07:30:00  H06     2   
...           ...        ...       ...  ...   ...   
14636    20101069 2019-05-25  21:40:00  J10     4   
14637    20102724 2019-05-26  19:50:00  I12     3   
14638    20100998 2019-05-25  20:30:00  H06     2   
14639    20091106 2019-05-17  20:15:00  F14     2   
14640    20106999 2019-05-30  21:35:00  H09     4   

                                          film slot type ticket type  \
0                            AVENGERS: HỒI KẾT       ĐÔI  Thành viên   
1                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
2                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
3                            AVENGERS: HỒI KẾT       ĐƠN  Thành viên   
4                            AVENGERS: HỒI KẾT       ĐÔI  Thành viên   
...                                        ...       ...         ...   
14636                   NGÔI ĐỀN KỲ QUÁI (C18)       ĐƠN  Thành viên   
14637                         ALADDIN (PD) (G)       ĐƠN  Thành viên   
14638                   NGÔI ĐỀN KỲ QUÁI (C18)       ĐÔI  Thành viên   
14639  JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)       ĐƠN  Thành viên   
14640  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)       ĐƠN  Thành viên   

       ticket price popcorn  saledate-showdate  ticketcount  
0            160000   Không               -1.0            1  
1             75000   Không               -1.0            1  
2             75000   Không               -1.0            3  
3             75000   Không               -1.0            2  
4            160000   Không               -1.0            1  
...             ...     ...                ...          ...  
14636         45000   Không               -1.0            3  
14637         45000   Không               -1.0            2  
14638         90000   Không               -1.0            1  
14639         45000   Không               -1.0            3  
14640         45000   Không               -1.0            2  

[14641 rows x 17 columns]

In [ ]:
df_ticket_distinct = df_ticket_distinct.drop(columns = ["ticketcode", "slot", "slot type", "ticket price", "saledate-showdate"])

In [ ]:
df_ticket_distinct

orderid  cashier   saledate     total  customerid  \
0                 10001052019B0003   emp008 2019-05-01  160000.0  0000024994   
1                 10001052019B0004   emp008 2019-05-01   75000.0  0000024994   
2                 10001052019B0005   emp008 2019-05-01  225000.0   KH6166700   
3                 10001052019B0006   emp008 2019-05-01  150000.0   KH6166700   
4                 10001052019B0007   emp008 2019-05-01  160000.0   KH6166700   
...                            ...      ...        ...       ...         ...   
14636  website00011910712019-05-25  Website 2019-05-25  135000.0  0001191071   
14637  website00011910712019-05-26  Website 2019-05-26   90000.0  0001191071   
14638  website00030041612019-05-25  Website 2019-05-25   90000.0  0003004161   
14639  website00030086172019-05-17  Website 2019-05-17  135000.0  0003008617   
14640  website00030086172019-05-30  Website 2019-05-30   90000.0  0003008617   

        showdate      time  room                                     film  \
0     2019-05-01  07:30:00     2                        AVENGERS: HỒI KẾT   
1     2019-05-01  07:30:00     2                        AVENGERS: HỒI KẾT   
2     2019-05-01  08:45:00     1                        AVENGERS: HỒI KẾT   
3     2019-05-01  07:30:00     2                        AVENGERS: HỒI KẾT   
4     2019-05-01  07:30:00     2                        AVENGERS: HỒI KẾT   
...          ...       ...   ...                                      ...   
14636 2019-05-25  21:40:00     4                   NGÔI ĐỀN KỲ QUÁI (C18)   
14637 2019-05-26  19:50:00     3                         ALADDIN (PD) (G)   
14638 2019-05-25  20:30:00     2                   NGÔI ĐỀN KỲ QUÁI (C18)   
14639 2019-05-17  20:15:00     2  JOHN WICK 3: CHUẨN BỊ CHIẾN TRANH (C18)   
14640 2019-05-30  21:35:00     4  CHÚA TỂ GODZILLA: ĐẾ VƯƠNG BẤT TỬ (C13)   

      ticket type popcorn  ticketcount  
0      Thành viên   Không            1  
1      Thành viên   Không            1  
2      Thành viên   Không            3  
3      Thành viên   Không            2  
4      Thành viên   Không            1  
...           ...     ...          ...  
14636  Thành viên   Không            3  
14637  Thành viên   Không            2  
14638  Thành viên   Không            1  
14639  Thành viên   Không            3  
14640  Thành viên   Không            2  

[14641 rows x 12 columns]

In [ ]:
df_ticket_distinct.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14641 entries, 0 to 14640
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   orderid      14641 non-null  object        
 1   cashier      14641 non-null  object        
 2   saledate     14641 non-null  datetime64[ns]
 3   total        14641 non-null  float64       
 4   customerid   14641 non-null  object        
 5   showdate     14641 non-null  datetime64[ns]
 6   time         14641 non-null  object        
 7   room         14641 non-null  int64         
 8   film         14641 non-null  object        
 9   ticket type  14641 non-null  object        
 10  popcorn      14641 non-null  object        
 11  ticketcount  14641 non-null  int64         
dtypes: datetime64[ns](2), float64(1), int64(2), object(7)
memory usage: 1.3+ MB


In [ ]:
df_ticket_distinct.to_excel("/Users/bethany.dinh/Desktop/Data Got Talent/ticket_cleanned_2_for_revenue.xlsx", engine='xlsxwriter')

### 3. Film vs Ticket

In [ ]:
df_film_added = pd.read_excel(f"Vòng 1-Working.xlsx", sheet_name = "film-added")